### PR TITLE
[feature](Nereids) support `select key encryptKey`

### PR DESCRIPTION
**Pull request by keanji-x**
## Proposed changes
add select key
```
- CREATE ENCRYPTKEY key_name AS "key_string"
- select key my_key
+-----------------------------+
| encryptKeyRef('', 'my_key') |
+-----------------------------+
| ABCD123456789               |
+-----------------------------+
```

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by XieJiann** [comment]
run buildall

**Comment by XieJiann** [comment]
run buildall

**Comment by hello-stephen** [comment]
(From new machine)TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 45.81 seconds
 stream load tsv:          541 seconds loaded 74807831229 Bytes, about 131 MB/s
 stream load json:         21 seconds loaded 2358488459 Bytes, about 107 MB/s
 stream load orc:          65 seconds loaded 1101869774 Bytes, about 16 MB/s
 stream load parquet:          31 seconds loaded 861443392 Bytes, about 26 MB/s
 insert into select:          29.3 seconds inserted 10000000 Rows, about 341K ops/s
 storage size: 17162071129 Bytes

File: fe/fe-core/src/main/antlr4/org/apache/doris/nereids/DorisParser.g4 [hunk 1]
```diff
@@ -373,6 +373,7 @@ primaryExpression
     | identifier                                                                               #columnReference
     | base=primaryExpression DOT fieldName=identifier                                          #dereference
     | LEFT_PAREN expression RIGHT_PAREN                                                        #parenthesizedExpression
+    | KEY (dbName=identifier DOT)? keyName=identifier                                                                   #encryptKey
```
Comment (morrySnow) [inline fe/fe-core/src/main/antlr4/org/apache/doris/nereids/DorisParser.g4:376]: align #xxxx

File: fe/fe-core/src/main/java/org/apache/doris/nereids/parser/LogicalPlanBuilder.java [hunk 2]
```diff
@@ -1101,6 +1101,17 @@ public UnboundFunction visitExtract(DorisParser.ExtractContext ctx) {
...
 
+    @Override
+    public Expression visitEncryptKey(DorisParser.EncryptKeyContext ctx) {
+        Expression db = new StringLiteral("");
+        if (ctx.dbName != null) {
+            db = new StringLiteral(ctx.dbName.getText());
+        }
+        Expression keyName = new StringLiteral(ctx.keyName.getText());
+        return new UnboundFunction("encryptKeyRef", false,
+                ImmutableList.of(db, keyName));
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/parser/LogicalPlanBuilder.java:1112]: return `new EncryptKeyRef(....);` is ok

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/expression/rules/FoldConstantRuleOnFE.java [hunk 3]
```diff
@@ -107,6 +113,23 @@ public Expression visitLiteral(Literal literal, ExpressionRewriteContext context
...
     }
 
+    @Override
+    public Expression visitEncryptKeyRef(EncryptKeyRef encryptKeyRef, ExpressionRewriteContext context) {
+        String dbName = encryptKeyRef.getDbName();
+        ConnectContext connectContext = ConnectContext.get();
+        if (Strings.isNullOrEmpty(dbName)) {
+            dbName = connectContext.getDatabase();
+        }
+        Preconditions.checkArgument(!"".equals(dbName), "DB " + dbName + "not found");
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/expression/rules/FoldConstantRuleOnFE.java:123]: throw `AnalysisException` directly

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/expression/rules/FoldConstantRuleOnFE.java [hunk 4]
```diff
@@ -107,6 +113,23 @@ public Expression visitLiteral(Literal literal, ExpressionRewriteContext context
...
+        dbName = ClusterNamespace.getFullName(connectContext.getClusterName(), dbName);
+        org.apache.doris.catalog.Database database =
+                Env.getCurrentEnv().getInternalCatalog().getDbNullable(dbName);
+        Preconditions.checkArgument(database != null, "DB " + dbName + "not found");
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/expression/rules/FoldConstantRuleOnFE.java:127]: throw AnalysisException directly

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/expression/rules/FoldConstantRuleOnFE.java [hunk 5]
```diff
@@ -107,6 +113,23 @@ public Expression visitLiteral(Literal literal, ExpressionRewriteContext context
...
+        EncryptKey encryptKey = database.getEncryptKey(encryptKeyRef.getEncryptKeyName());
+        Preconditions.checkArgument(encryptKey != null, "Can not found encryptKey" + encryptKeyRef.getEncryptKeyName());
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/expression/rules/FoldConstantRuleOnFE.java:129]: throw AnalysisException directly

File: fe/fe-core/src/main/java/org/apache/doris/catalog/BuiltinScalarFunctions.java [hunk 6]
```diff
@@ -463,6 +464,7 @@ public class BuiltinScalarFunctions implements FunctionHelper {
             scalar(ElementExtract.class, "%element_extract%"),
             scalar(ElementSlice.class, "%element_slice%"),
             scalar(Elt.class, "elt"),
+            scalar(EncryptKeyRef.class, "encryptKeyRef"),
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/catalog/BuiltinScalarFunctions.java:467]: should not put this function in function list

**Comment by XieJiann** [comment]
run buildall

**Comment by hello-stephen** [comment]
(From new machine)TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 45.32 seconds
 stream load tsv:          545 seconds loaded 74807831229 Bytes, about 130 MB/s
 stream load json:         20 seconds loaded 2358488459 Bytes, about 112 MB/s
 stream load orc:          65 seconds loaded 1101869774 Bytes, about 16 MB/s
 stream load parquet:          31 seconds loaded 861443392 Bytes, about 26 MB/s
 insert into select:          29.4 seconds inserted 10000000 Rows, about 340K ops/s
 storage size: 17162084846 Bytes

**Comment by XieJiann** [comment]
run buildall

**Review by morrySnow — DISMISSED** [review]

**Comment by XieJiann** [comment]
run buildall

**Comment by hello-stephen** [comment]
(From new machine)TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 47.43 seconds
 stream load tsv:          538 seconds loaded 74807831229 Bytes, about 132 MB/s
 stream load json:         20 seconds loaded 2358488459 Bytes, about 112 MB/s
 stream load orc:          66 seconds loaded 1101869774 Bytes, about 15 MB/s
 stream load parquet:          31 seconds loaded 861443392 Bytes, about 26 MB/s
 insert into select:          29.5 seconds inserted 10000000 Rows, about 338K ops/s
 storage size: 17161873934 Bytes

**Review by morrySnow — APPROVED** [review]

**Review by jackwener — APPROVED** [review]

**Pull request merged**